# Analysing CIA Factbook Data Using SQL

## Introduction

In this project, we'll work with data from the CIA World Factbook, a compendium of statistics about all of the countries on Earth. The Factbook contains demographic information like the following:

- **population** — the global population.
- **population_growth** — the annual population growth rate, as a percentage.
- **area** — the total land and water area.

First, we must connect out Jupyter Notebook to our database file via the following chunk of code:

In [1]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

'Connected: None@factbook.db'

## Overview of the Data

Let's start off by writing a query to return information on the tables within the database.

In [2]:
%%sql
SELECT *
    FROM sqlite_master
    WHERE type='table';

Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

Now, we'll write and run a query that returns the first five rows of the **facts** table.

In [3]:
%%sql
SELECT * FROM facts
LIMIT 5;

Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

## Summary Statistics

To further explore the data, we'll calculate some summary statistics and look for any countries with outlier values relating to population.

In [4]:
%%sql
SELECT MIN(population) AS min_pop,
    MAX(population) AS max_pop,
    MIN(population_growth) AS min_growth,
    MAX(population_growth) AS max_growth
FROM facts;

Done.


[(0, 7256490011, 0.0, 4.02)]

## Exploring Outliers

We see a few interesting things in the summary statistics on the previous screen:

- There's a country with a population of 0
- There's a country with a population of 7256490011 (or **more than 7.2 billion people**)
- Let's use subqueries to zoom in on just these countries without using the specific values.

Let's find the countries with the minimum population

In [5]:
%%sql
SELECT name FROM facts
WHERE population = (
    SELECT MIN(population)
    FROM facts
);

Done.


[('Antarctica',)]

Now, we'll find the ones with the maximum population

In [6]:
%%sql
SELECT name FROM facts
WHERE population = (
    SELECT MAX(population)
    FROM facts
);

Done.


[('World',)]

## Exploring Population and Area

### Recalculate Population Summary Statistics

It seems like the table contains a row for the whole world, which explains the population of over 7.2 billion. It also seems like the table contains a row for Antarctica, which explains the population of 0. This seems to match the CIA Factbook page for Antarctica.

Now that we know this, we should recalculate the summary statistics we calculated earlier — this time excluding the row for the whole world.

In [7]:
%%sql
SELECT MIN(population) AS min_pop,
    MAX(population) AS max_pop,
    MIN(population_growth) AS min_growth,
    MAX(population_growth) AS max_growth
FROM facts
WHERE name <> 'World';

Done.


[(0, 1367485388, 0.0, 4.02)]

### Calculate Average Population and Area

The average value for population is:

In [9]:
%%sql
SELECT AVG(population) AS avg_pop
FROM facts;

Done.


[(62094928.32231405,)]

And the average value for area is:

In [10]:
%%sql
SELECT AVG(area) AS avg_area
FROM facts;

Done.


[(555093.546184739,)]

## Finding Densely Populated Countries

To finish, we'll build on the query we wrote for the previous screen to find countries that are densely populated. We'll identify countries that have the following:

- Above-average values for population
- Below-average values for area

In [11]:
%%sql
SELECT name FROM facts
WHERE population > (
    SELECT AVG(population)
    FROM facts
);

Done.


[('Bangladesh',),
 ('Brazil',),
 ('China',),
 ('Congo, Democratic Republic of the',),
 ('Egypt',),
 ('Ethiopia',),
 ('France',),
 ('Germany',),
 ('India',),
 ('Indonesia',),
 ('Iran',),
 ('Japan',),
 ('Mexico',),
 ('Nigeria',),
 ('Pakistan',),
 ('Philippines',),
 ('Russia',),
 ('Thailand',),
 ('Turkey',),
 ('United Kingdom',),
 ('United States',),
 ('Vietnam',),
 ('European Union',),
 ('World',)]

In [12]:
%%sql
SELECT name FROM facts
WHERE area < (
    SELECT AVG(area)
    FROM facts
);

Done.


[('Albania',),
 ('Andorra',),
 ('Antigua and Barbuda',),
 ('Armenia',),
 ('Austria',),
 ('Azerbaijan',),
 ('Bahamas, The',),
 ('Bahrain',),
 ('Bangladesh',),
 ('Barbados',),
 ('Belarus',),
 ('Belgium',),
 ('Belize',),
 ('Benin',),
 ('Bhutan',),
 ('Bosnia and Herzegovina',),
 ('Brunei',),
 ('Bulgaria',),
 ('Burkina Faso',),
 ('Burundi',),
 ('Cambodia',),
 ('Cameroon',),
 ('Cabo Verde',),
 ('Comoros',),
 ('Congo, Republic of the',),
 ('Costa Rica',),
 ("Cote d'Ivoire",),
 ('Croatia',),
 ('Cuba',),
 ('Cyprus',),
 ('Czech Republic',),
 ('Denmark',),
 ('Djibouti',),
 ('Dominica',),
 ('Dominican Republic',),
 ('Ecuador',),
 ('El Salvador',),
 ('Equatorial Guinea',),
 ('Eritrea',),
 ('Estonia',),
 ('Fiji',),
 ('Finland',),
 ('Gabon',),
 ('Gambia, The',),
 ('Georgia',),
 ('Germany',),
 ('Ghana',),
 ('Greece',),
 ('Grenada',),
 ('Guatemala',),
 ('Guinea',),
 ('Guinea-Bissau',),
 ('Guyana',),
 ('Haiti',),
 ('Honduras',),
 ('Hungary',),
 ('Iceland',),
 ('Iraq',),
 ('Ireland',),
 ('Israel',),
 ('Italy',),
 ('Jamaica',),
 ('Japan',),
 ('Jordan',),
 ('Kiribati',),
 ('Korea, North',),
 ('Korea, South',),
 ('Kosovo',),
 ('Kuwait',),
 ('Kyrgyzstan',),
 ('Laos',),
 ('Latvia',),
 ('Lebanon',),
 ('Lesotho',),
 ('Liberia',),
 ('Liechtenstein',),
 ('Lithuania',),
 ('Luxembourg',),
 ('Macedonia',),
 ('Malawi',),
 ('Malaysia',),
 ('Maldives',),
 ('Malta',),
 ('Marshall Islands',),
 ('Mauritius',),
 ('Micronesia, Federated States of',),
 ('Moldova',),
 ('Monaco',),
 ('Montenegro',),
 ('Morocco',),
 ('Nauru',),
 ('Nepal',),
 ('Netherlands',),
 ('New Zealand',),
 ('Nicaragua',),
 ('Norway',),
 ('Oman',),
 ('Palau',),
 ('Panama',),
 ('Papua New Guinea',),
 ('Paraguay',),
 ('Philippines',),
 ('Poland',),
 ('Portugal',),
 ('Qatar',),
 ('Romania',),
 ('Rwanda',),
 ('Saint Kitts and Nevis',),
 ('Saint Lucia',),
 ('Saint Vincent and the Grenadines',),
 ('Samoa',),
 ('San Marino',),
 ('Sao Tome and Principe',),
 ('Senegal',),
 ('Serbia',),
 ('Seychelles',),
 ('Sierra Leone',),
 ('Singapore',),
 ('Slovakia',),
 ('Slovenia',),
 ('Solomon Islands',),
 ('Spain',),
 ('Sri Lanka',),
 ('Suriname',),
 ('Swaziland',),
 ('Sweden',),
 ('Switzerland',),
 ('Syria',),
 ('Tajikistan',),
 ('Thailand',),
 ('Timor-Leste',),
 ('Togo',),
 ('Tonga',),
 ('Trinidad and Tobago',),
 ('Tunisia',),
 ('Turkmenistan',),
 ('Tuvalu',),
 ('Uganda',),
 ('United Arab Emirates',),
 ('United Kingdom',),
 ('Uruguay',),
 ('Uzbekistan',),
 ('Vanuatu',),
 ('Holy See (Vatican City)',),
 ('Vietnam',),
 ('Yemen',),
 ('Zimbabwe',),
 ('Taiwan',),
 ('Ashmore and Cartier Islands',),
 ('Christmas Island',),
 ('Cocos (Keeling) Islands',),
 ('Coral Sea Islands',),
 ('Heard Island and McDonald Islands',),
 ('Norfolk Island',),
 ('Hong Kong',),
 ('Macau',),
 ('Faroe Islands',),
 ('Clipperton Island',),
 ('French Polynesia',),
 ('New Caledonia',),
 ('Saint Martin',),
 ('Saint Pierre and Miquelon',),
 ('Wallis and Futuna',),
 ('Aruba',),
 ('Curacao',),
 ('Sint Maarten',),
 ('Cook Islands',),
 ('Niue',),
 ('Tokelau',),
 ('Bouvet Island',),
 ('Jan Mayen',),
 ('Svalbard',),
 ('Akrotiri',),
 ('Anguilla',),
 ('Bermuda',),
 ('British Indian Ocean Territory',),
 ('British Virgin Islands',),
 ('Cayman Islands',),
 ('Dhekelia',),
 ('Falkland Islands (Islas Malvinas)',),
 ('Gibraltar',),
 ('Guernsey',),
 ('Jersey',),
 ('Isle of Man',),
 ('Montserrat',),
 ('Pitcairn Islands',),
 ('Saint Helena, Ascension, and Tristan da Cunha',),
 ('South Georgia and South Sandwich Islands',),
 ('Turks and Caicos Islands',),
 ('American Samoa',),
 ('Guam',),
 ('Navassa Island',),
 ('Northern Mariana Islands',),
 ('Puerto Rico',),
 ('Virgin Islands',),
 ('Wake Island',),
 ('Gaza Strip',),
 ('Paracel Islands',),
 ('Spratly Islands',),
 ('West Bank',),
 ('Western Sahara',)]